stdlib/regex.gr

```diff
@@ -0,0 +1,2983 @@
+    if (pos == len) {
+      consRange(since, pos, [])
+    } else if (replacementExploded[pos] == '$') {
+      let c = if ((pos + 1) < len) Some(replacementExploded[pos + 1]) else None
+      if (c == Some('&')) {
+        consRange(since, pos, [getInputSubstr(0), ...loop(pos + 2, pos + 2)])
+      } else if (c == Some('`')) {
+        consRange(since, pos, [getBeforeMatch(), ...loop(pos + 2, pos + 2)])
+      } else if (c == Some('\'')) {
+        consRange(since, pos, [getAfterMatch(), ...loop(pos + 2, pos + 2)])
+      } else {
+        consRange(since, pos, {
+          if (c == Some('$')) {
+            loop(pos + 2, pos + 1)
+          } else if (c == Some('.')) {
+            loop(pos + 2, pos + 2)
+          } else {
+            let rec dLoop = (pos, accum) => {
+              if (pos == len) {
+                [getInputSubstr(accum)]
+              } else {
+                let c = replacementExploded[pos]
+                if (Char.code('0') <= Char.code(c) && Char.code(c) <= Char.code('9')) {
+                  dLoop(pos + 1, (10 * accum) + (Char.code(c) - Char.code('0')))
+                } else {
+                  [getInputSubstr(accum), ...loop(pos, pos)]
+                }
+              }
+            }
+            dLoop(pos + 1, 0)
+          }
+        })
+      }
+    } else {
+      loop(pos + 1, since)
+    }
+  }
+  let res = loop(0, 0)
+  List.reduceRight(String.concat, "", res)
+}
+
+
+let regexReplaceHelp = (rx: RegularExpression, toSearch: String, replacement: String, all: Bool) => {
+  let buf = makeMatchBuffer(toSearch)
+  let mut out = []
+  let rec loop = (searchPos) => {
+    let state = Array.make(rx.reNumGroups, None)
+    let poss = searchMatch(rx, buf, searchPos, searchPos, Array.length(buf.matchInputExploded), state)
+    let recur = (start, end) => {
+      if (end == searchPos) {
+        if (searchPos == String.length(toSearch)) {
+          ""
+        } else {
+          String.concat(String.slice(searchPos, searchPos + 1, toSearch), loop(searchPos + 1))
+        }
+      } else {
+        loop(end)
+      }
+    }
+    match(poss) {
+      None => if (searchPos == 0) toSearch else String.slice(searchPos, String.length(toSearch), toSearch),
+      Some((start, end)) =>
+        String.concat(String.slice(searchPos, start, toSearch),
+          String.concat(computeReplacement(buf, replacement, start, end, state),
+            if (all) recur(start, end) else String.slice(end, String.length(toSearch), toSearch))),
+    }
+  }
+  loop(0)
+}
+
+/**
+ * Replaces the first match for the given regular expression contained within the given string with the given replacement.
+ * Replacement strings support the following syntax:
+ * - `$&` - Replaced with the text of the matching portion of input (e.g. for `(foo)`, the search string `foo bar`, and the replacement `baz $&`, the result will be `baz foo bar`)
+ * - `$n` / `$nn` (where `n` is a digit) - Replaced with the text of group `nn`
```

Comment:
why do we have `nn`? is that a double digit number? Do we not support triple digit?

Reply:
I don't know any regex engine that supports 3 numbers. That'd be quite the thicc regex!

Reply:
Sounds good. I notice the MDN docs say "where n under 100" and I think our docs are better 👍